atom/src/validatebehavior.cpp

```diff
@@ -278,33 +282,58 @@ float_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject
     return validate_type_fail( member, atom, newvalue, "float" );
 }
 
+static PyObject*
+bytes_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalue )
+{
+    if( Py23Bytes_Check( newvalue ) )
+        return newref( newvalue );
+    return validate_type_fail( member, atom, newvalue, "bytes" );
+}
+
+static PyObject*
+bytes_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalue )
+{
+    if( Py23Bytes_Check( newvalue ) )
+        return newref( newvalue );
+
+    if( PyUnicode_Check( newvalue ) )
+        return PyUnicode_AsUTF8String( newvalue );
+
+    return validate_type_fail( member, atom, newvalue, "bytes" );
+}
 
 static PyObject*
-str_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalue )
+string_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalue )
 {
-    if( PyString_Check( newvalue ) )
+    if( Py23Str_Check( newvalue ) )
         return newref( newvalue );
     return validate_type_fail( member, atom, newvalue, "str" );
 }
 
 
 static PyObject*
-str_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalue )
+string_promote_handler( Member* member, CAtom* atom, PyObject* oldvalue, PyObject* newvalue )
 {
-    if( PyString_Check( newvalue ) )
+    if( Py23Str_Check( newvalue ) )
         return newref( newvalue );
+#if PY_MAJOR_VERSION >= 3
+    if( PyBytes_Check( newvalue ) )
+        return PyUnicode_FromString( PyBytes_AS_STRING( newvalue ) );
+
+#else
     if( PyUnicode_Check( newvalue ) )
         return PyUnicode_AsUTF8String( newvalue );
+
+#endif
```

Comment:
Not quite, no. The docstring for `Str` says:

>Under Python 2 this is a byte string, under Python 3 a unicode one.
>
>The use of this member is discouraged in Python 2/3 compatible codebase
>as Bytes and Unicode provide a more homogeneous behavior.

The other types describe what their promotion semantics are.

The behavior I see here:

INPUT | Unicode | Bytes
----- | ------- | -----
Py2 |  UTF-8 bytestring | Bytes
Py3 | Unicode | Unicode